### DISCUSSION TITLE
Database drift detected when changing foreign key from required to nullable

### DISCUSSION THREAD

**Discussion by camsloanftc** [prisma/prisma#19546] [Q&A]
Scenario:
I am trying to change a model that previously had a required foreign userId key, to making that key optional.

```prisma
model Token {
    // ... other fields
    
    // previously
    // user        User     @relation(fields: [userId], references: [id])
    // userId      Int

    // new
    // user        User?     @relation(fields: [userId], references: [id])
    // userId      Int?

}
```

I was doing prisma db push while developing, and now when I try to run prisma migrate dev, I get:

```
Drift detected: Your database schema is not in sync with your migration history.

The following is a summary of the differences between the expected database schema given your migrations files, and the actual schema of the database.

It should be understood as the set of changes to get from the expected schema to the actual schema.

[*] Changed the `PlanName` enum
  [+] Added variant `APP_SUMO_TIER_1`
  [+] Added variant `APP_SUMO_TIER_2`
  [+] Added variant `APP_SUMO_TIER_3`
  [+] Added variant `APP_SUMO_TIER_4`
  [+] Added variant `EARLY_STAGE`

[*] Changed the `TokenType` enum
  [+] Added variant `SET_PASSWORD`

[*] Changed the `Company` table
  [+] Added column `appSumoInvoiceItemId`
  [+] Added column `appSumoLicenseId`

[*] Changed the `Token` table
  [-] Removed foreign key on columns (userId)
  [+] Added column `data`
  [*] Altered column `sentTo` (changed from Required to Nullable)
  [*] Altered column `userId` (changed from Required to Nullable)
  [+] Added foreign key on columns (userId)

✔ We need to reset the "public" schema at "localhost:32768"
Do you want to continue? All data will be lost. … no
```

You can see the other changes there, but they are all additive. For some reason, prisma is trying to remove the foreign key, and then re-add it, which I believe is why it is trying to reset the DB.

To be honest, I don’t mind resetting the local DB, but will this migration run successfully when it hits production? That would be a big concern for me.

Also, another concern based on the diff: is the migration going to drop the foreign key column and then add a new foreign key column? Would that populate existing entries or they would be orphaned?

**Comment by ludralph** [comment]
Hi @camsloanftc 👋 

> Will the migration drop the foreign key column and then add a new foreign key column? Will existing entries be orphaned?

Based on the information provided, it seems that the migration is removing and re-adding the foreign key column. When the foreign key column is dropped, the existing entries in the Token table that were associated with a User will become orphaned, meaning they will no longer have a valid reference to a User entry. In some cases, the database may prevent the removal of the foreign key if there are existing references, resulting in an error during migration. In other cases, the database may allow the removal, and the orphaned entries will remain in the Token table with a null value in the userId column.

> To be honest, I don’t mind resetting the local DB, but will this migration run successfully when it hits production? That would be a big concern for me.

Before applying the migration in the production environment, set up a staging environment that closely resembles your production setup. Apply the migration to the staging environment and thoroughly test your application to ensure that it works as expected and there are no data integrity issues. This will give you confidence in running the migration in the production environment.

**Reply by nurul3101** [reply]
@camsloanftc Did you had a chance to look at @ludralph's response?

It would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by nurul3101** [reply]
We are closing this Discussion due to inactivity. Please open a new Discussion if you have a similar problem 🙏
